### PR TITLE
docs: undocument private constructors

### PR DESCRIPTION
**Pull request by SpaceEEC**
**Please describe the changes this PR makes and why it should be merged:**

This PR removes documentation for private constructors.
Those being documented confused library consumers into thinking that they are supposed / intended to instantiate those classes themselves.

Some constructors were intended to be documented due to `Structures.extend` use-cases, but this feature has, since then, been removed.

Changed classes:
- `Shard`
- `ShardClientUtil`
- `BaseGuildTextChannel`
- `ClientPresence`
- `DMChannel`
- `GuildBan`
- `GuildChannel`
- `GuildEmoji`
- `GuildTemplate`
- `Message`
- `MessageReaction`
- `Presence`
- `Role`
- `Sticker`
- `StickerPack`
- `StoreChannel`
- `ThreadChannel`
- `ThreadMember`
- `Typing`
- `User`
- `VoiceState`
- `Widget`
- `WidgetMember`

This PR also modifies the types to no longer allow instantiation where unintended, this includes few more classes than listed above.

**Status and versioning classification:**
- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

File: src/structures/GuildBan.js [hunk 1]
```diff
@@ -7,11 +7,6 @@ const Base = require('./Base');
  * @extends {Base}
  */
 class GuildBan extends Base {
-  /**
-   * @param {Client} client The instantiating client
-   * @param {APIBanData} data The data for the ban
```
Comment (Jiralite) [inline src/structures/GuildBan.js:12]: This was the last instance of `APIBanData`. Should we remove this too?

https://github.com/discordjs/discord.js/blob/395186da7241b1ea704e4d5e0c6df6fd4655eb5c/src/structures/GuildBan.js#L59-L62

File: src/structures/GuildTemplate.js [hunk 2]
```diff
@@ -9,10 +9,6 @@ const DataResolver = require('../util/DataResolver');
  * @extends {Base}
  */
 class GuildTemplate extends Base {
-  /**
-   * @param {Client} client The instantiating client
-   * @param {APIGuildTemplate} data The raw data for the template
```
Comment (Jiralite) [inline src/structures/GuildTemplate.js:14]: With this removal, the only other instance of `APIGuildTemplate` is in the `_patch()` method below. Does this really need to be documented? I noticed other `_patch()` methods aren't documented so if both are to go, we could remove `@external` altogether!

File: src/structures/MessageReaction.js [hunk 3]
```diff
@@ -9,11 +9,6 @@ const Util = require('../util/Util');
  * Represents a reaction to a message.
  */
 class MessageReaction {
-  /**
-   * @param {Client} client The instantiating client
-   * @param {APIReaction} data The data for the message reaction
```
Comment (Jiralite) [inline src/structures/MessageReaction.js:14]: This was the last instance of `APIReaction`. Should we remove this too?

https://github.com/discordjs/discord.js/blob/395186da7241b1ea704e4d5e0c6df6fd4655eb5c/src/structures/MessageReaction.js#L134-L137

File: src/structures/ClientPresence.js [hunk 4]
```diff
@@ -9,10 +9,6 @@ const { ActivityTypes, Opcodes } = require('../util/Constants');
  * @extends {Presence}
  */
 class ClientPresence extends Presence {
-  /**
-   * @param {Client} client The instantiating client
-   * @param {APIPresence} [data={}] The data for the client presence
```
Comment (Jiralite) [inline src/structures/ClientPresence.js:14]: If `APIPresence` doesn't need to be documented in the lone private method `_parse()` below, the `@external` could be removed too?

File: src/structures/Sticker.js [hunk 5]
```diff
@@ -9,10 +9,6 @@ const SnowflakeUtil = require('../util/SnowflakeUtil');
  * @extends {Base}
  */
 class Sticker extends Base {
-  /**
-   * @param {Client} client The instantiating client
-   * @param {APISticker | APIStickerItem} sticker The data for the sticker
```
Comment (Jiralite) [inline src/structures/Sticker.js:14]: This was the last instance of `APIStickerItem`. Should we remove this too?
https://github.com/discordjs/discord.js/blob/395186da7241b1ea704e4d5e0c6df6fd4655eb5c/src/structures/Sticker.js#L238-L241

File: src/structures/StickerPack.js [hunk 6]
```diff
@@ -10,10 +10,6 @@ const SnowflakeUtil = require('../util/SnowflakeUtil');
  * @extends {Base}
  */
 class StickerPack extends Base {
-  /**
-   * @param {Client} client The instantiating client
-   * @param {APIStickerPack} pack The data for the sticker pack
```
Comment (Jiralite) [inline src/structures/StickerPack.js:15]: This was the last instance of `APIStickerPack`. Should we remove this too?

https://github.com/discordjs/discord.js/blob/395186da7241b1ea704e4d5e0c6df6fd4655eb5c/src/structures/StickerPack.js#L97-L100

File: src/structures/ThreadMember.js [hunk 7]
```diff
@@ -8,10 +8,6 @@ const ThreadMemberFlags = require('../util/ThreadMemberFlags');
  * @extends {Base}
  */
 class ThreadMember extends Base {
-  /**
-   * @param {ThreadChannel} thread The thread that this member is associated with
-   * @param {APIThreadMember} data The data for the thread member
```
Comment (Jiralite) [inline src/structures/ThreadMember.js:13]: This was the last instance of `APIThreadMember`. Should we remove this too?

https://github.com/discordjs/discord.js/blob/395186da7241b1ea704e4d5e0c6df6fd4655eb5c/src/structures/ThreadMember.js#L94-L97

File: src/structures/Typing.js [hunk 8]
```diff
@@ -7,11 +7,6 @@ const Base = require('./Base');
  * @extends {Base}
  */
 class Typing extends Base {
-  /**
-   * @param {TextBasedChannels} channel The channel this typing came from
-   * @param {User} user The user that started typing
-   * @param {APITypingStart} data The raw data received
```
Comment (Jiralite) [inline src/structures/Typing.js:13]: This was the last instance of `APITypingStart`. Should we remove this too?

https://github.com/discordjs/discord.js/blob/395186da7241b1ea704e4d5e0c6df6fd4655eb5c/src/structures/Typing.js#L74-L77

File: src/structures/VoiceState.js [hunk 9]
```diff
@@ -7,10 +7,6 @@ const { Error, TypeError } = require('../errors');
  * Represents the voice state for a Guild Member.
  */
 class VoiceState extends Base {
-  /**
-   * @param {Guild} guild The guild the voice state is part of
-   * @param {APIVoiceState} data The data for the voice state
```
Comment (Jiralite) [inline src/structures/VoiceState.js:12]: This was the last instance of `APIVoiceState`. Should we remove this too?

https://github.com/discordjs/discord.js/blob/395186da7241b1ea704e4d5e0c6df6fd4655eb5c/src/structures/VoiceState.js#L229-L232

File: src/structures/ClientPresence.js (hunk 4 above)
Comment (SpaceEEC) [reply src/structures/ClientPresence.js:14]: I think that docstring provides some value, so I'm opting to keep it.
I removed all the other things as per your suggestions however, thank you for these!

**Review by Jiralite — APPROVED** [review]

**Review by ImRodry — APPROVED** [review]
Was gonna submit the review about the external links as soon as Jiralite did it so LGTM xD

**Review by ImRodry — CHANGES_REQUESTED** [review]
Actually, most of these constructors are public in the typings, could you go through them and make them private?

**Review by iCrawl — APPROVED** [review]

**Review by vladfrangu — APPROVED** [review]

**Comment by tremby** [comment]
Is there a way, short of `@ts-expect-error`, I can opt out of this for my tests? I use these constructors to produce fixtures for the purpose of unit testing my application.

**Pull request merged**